File: src/types/book.ts
export type BookChapter = {
  id: string
  bookId?: string
  parentId?: string | null
  title: string
  order?: number
  level?: number
  summary?: string | null
  markdownPath?: string
  createdAt?: string
  updatedAt?: string
  children?: BookChapter[]
}

export type BookListItem = {
  id: string
  knowledgeBaseId?: string
  title: string
  author?: string | null
  dynasty?: string | null
  description?: string | null
  coverUrl?: string | null
  chapterCount?: number
  createdAt?: string
  updatedAt?: string
  chapters: BookChapter[]
}

export type BookListResponse = BookListItem[]

export type BookAnchor = {
  id: string
  chapterId: string
  title: string
  startOffset: number
  endOffset: number
  markdownHeading?: string | null
}

export type BookEntityMention = {
  id: string
  entityId: string
  label: string
  nodeType: string
  chapterId: string
  anchorId?: string | null
  startOffset: number
  endOffset: number
}

export type BookTextResponse = {
  chapterId?: string
  bookId: string
  title: string
  markdown: string
  anchors: BookAnchor[]
  entities: BookEntityMention[]
  content: string
}

export type BookSelectedText = {
  chapterId: string
  text: string
  startOffset: number
  endOffset: number
}

export type BookChatScope = 'chapter' | 'book'

export type BookChatRequest = {
  sessionId: string
  knowledgeBaseId?: string
  bookId?: string
  chapterId?: string
  chatScope?: BookChatScope
  question: string
  selectedText?: BookSelectedText
}

export type BookSourceRef = {
  id: string
  knowledgeBaseId?: string
  bookId: string
  bookTitle?: string
  chapterId: string
  chapterTitle?: string
  anchorId?: string | null
  title: string
  quote: string
  startOffset: number
  endOffset: number
}

export type BookChatResponse = {
  answer: string
  sourceRefs: BookSourceRef[]
  relatedNodeIds: string[]
}

File: app/models/book.py
from sqlalchemy import Float, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Book(Base):
    __tablename__ = "books"

    id: Mapped[str] = mapped_column(String(80), primary_key=True)
    knowledge_base_id: Mapped[str] = mapped_column(String(80), index=True)
    title: Mapped[str] = mapped_column(String(200))
    author: Mapped[str | None] = mapped_column(String(200), nullable=True)
    dynasty: Mapped[str | None] = mapped_column(String(100), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    cover_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    chapter_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[str] = mapped_column(String(40))
    updated_at: Mapped[str] = mapped_column(String(40))


class Chapter(Base):
    __tablename__ = "chapters"

    id: Mapped[str] = mapped_column(String(80), primary_key=True)
    book_id: Mapped[str] = mapped_column(ForeignKey("books.id"), index=True)
    parent_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    title: Mapped[str] = mapped_column(String(200))
    sort_order: Mapped[int] = mapped_column("order", Integer)
    level: Mapped[int] = mapped_column(Integer)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    markdown_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[str] = mapped_column(String(40))
    updated_at: Mapped[str] = mapped_column(String(40))


class ChapterContent(Base):
    __tablename__ = "chapter_contents"

    chapter_id: Mapped[str] = mapped_column(
        ForeignKey("chapters.id"),
        primary_key=True,
    )
    book_id: Mapped[str] = mapped_column(ForeignKey("books.id"), index=True)
    title: Mapped[str] = mapped_column(String(200))
    markdown: Mapped[str] = mapped_column(Text)
    anchors: Mapped[list] = mapped_column(JSON, default=list)
    entities: Mapped[list] = mapped_column(JSON, default=list)


class GraphNode(Base):
    __tablename__ = "graph_nodes"

    id: Mapped[str] = mapped_column(String(80), primary_key=True)
    label: Mapped[str] = mapped_column(String(200))
    node_type: Mapped[str] = mapped_column(String(80), index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)


class GraphEdge(Base):
    __tablename__ = "graph_edges"

    id: Mapped[str] = mapped_column(String(80), primary_key=True)
    source: Mapped[str] = mapped_column(String(80), index=True)
    target: Mapped[str] = mapped_column(String(80), index=True)
    label: Mapped[str] = mapped_column(String(200))
    weight: Mapped[float | None] = mapped_column(Float, nullable=True)


class SourceRef(Base):
    __tablename__ = "source_refs"

    id: Mapped[str] = mapped_column(String(80), primary_key=True)
    knowledge_base_id: Mapped[str] = mapped_column(String(80), index=True)
    book_id: Mapped[str] = mapped_column(ForeignKey("books.id"), index=True)
    chapter_id: Mapped[str] = mapped_column(ForeignKey("chapters.id"), index=True)
    anchor_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    node_id: Mapped[str | None] = mapped_column(String(80), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(200))
    quote: Mapped[str | None] = mapped_column(Text, nullable=True)
    start_offset: Mapped[int | None] = mapped_column(Integer, nullable=True)
    end_offset: Mapped[int | None] = mapped_column(Integer, nullable=True)
